content/advent-2016/machine-learning-google-prediction-api.md

```diff
@@ -0,0 +1,224 @@
+            "Please predict which language this is"
+        ]
+    }
+}
+```
+
+Yes, it is a little weird that we're embedding CSV into JSON, and I expect this will evolve in future versions
+of the API. But it is still pretty clear what's going on; we're asking the Google Prediction API to predict which
+language the `"Please predict which language this is"` sentence is in.
+
+## Building a little app in Go
+
+We will write a little command line tool that lets us query the model from the terminal. 
+
+As the example project for this article was being put together, it became clear that the lion share of the work
+was actually in authorising the requests via OAuth. And the experience for a terminal app was less than desirable; you 
+had to copy and paste a URL into a browser, log in (or create an account), click Accept to allow access to the API, 
+then copy the access code from the query parameters of a failed redirect and paste it into the terminal.
+
+OAuth is a web based authorisation protocol through and through, and short of implementing an HTTP endpoint to
+receive the request, there isn't much we can do to make the experience any better. So instead, we are going to be
+a little cheeky and piggy back on [Google's Try it app hosted on App Engine](http://try-prediction.appspot.com).
+
+In your `$GOPATH/src` folder, create a new folder called `language-predict` and add a `main.go` file.
+
+We will write a program that reads a line from standard input, queries the model, and writes the response in a nice
+human readable way.
+
+Add the following code to `main.go`:
+
+```go
+var labels = map[string]string{
+	"English": "🇬🇧 English",
+	"Spanish": "🇪🇸 Spanish",
+	"French":  "🇫🇷 French",
+}
+
+func main() {
+	s := bufio.NewScanner(os.Stdin)
+	fmt.Print("> ")
+	for s.Scan() {
+		res, err := do(s.Text())
+		if err != nil {
+			fmt.Println("failed:", err)
+			continue
+		}
+		label, ok := labels[res.OutputLabel]
+		if !ok {
+			label = res.OutputLabel
+		}
+		fmt.Printf("  ^^ That looks like %s to me\n", label)
+		fmt.Print("> ")
+	}
+}
+```
+
+* You can copy and paste the flags from here if you like :)
+
+This code uses a `bufio.Scanner` to read a line at a time from `os.Stdin`, which allows us to type a sentence, and
+hit return which will unblock the `Scan` method and execute the `for` block.
+
+We then ask the `do` function to do its magic getting the text string via the `Text()` method on the `Scanner`,
+and attempt to add a nice label (that includes a flag) to the result, before writing it out using `fmt.Print` calls.
+
+The response we expect back from the Google Prediction API is modelled in our `response` struct that we will 
+add next:
+
+```go
+type response struct {
+	Kind        string
+	ID          string
+	OutputLabel string
+	OutputMulti []struct {
+		Label string
+		Score string
+	}
+	OutputValue float64
+}
+```
+
+The API gives us more information than we need, including a confidence level for each possible category. So for
```

Comment:
Perhaps you should tie in the previous mentions of regression vs categorisation here? Eg `including a confidence level as this is a regression not categorisation`?